packages/vite/rollup.dts.config.ts

```diff
@@ -72,13 +74,64 @@ function patchTypes(): Plugin {
         }
       }
 
+      // Rollup deduplicate type names with a trailing `$1` or `$2`, which can be
+      // confusing when showed in autocompletions. Try to replace with a better name
+      const foundDollarNames = new Set<string>()
+      for (const match of code.matchAll(identifierWithTrailingDollarRE)) {
+        foundDollarNames.add(match[0])
+      }
+      for (const name of foundDollarNames) {
+        const betterName = getBetterTypeName(name)
+        if (!betterName) {
+          this.warn(
+            `${chunk.fileName} contains "${name}" which is a confusing type name`,
+          )
+          process.exitCode = 1
+          continue
+        }
+        const regexEscapedName = escapeRegex(name)
+        // If the better name accesses a namespace, the existing `Foo as Foo$1`
+        // named import cannot be replaced with `Foo as Namespace.Foo`, so we
+        // pre-emptively remove the whole named import
+        if (betterName.includes('.')) {
+          code = code.replace(
+            new RegExp(`\\b\\w+\\b as ${regexEscapedName},?\\s?`),
+            '',
+          )
+        }
+        code = code.replace(
+          new RegExp(`\\b${regexEscapedName}\\b`, 'g'),
+          betterName,
+        )
+      }
+
       // Clean unnecessary comments
-      return code
+      code = code
         .replace(singlelineCommentsRE, '')
         .replace(multilineCommentsRE, (m) => {
           return licenseCommentsRE.test(m) ? '' : m
         })
         .replace(consecutiveNewlinesRE, '\n\n')
+
+      return code
     },
   }
 }
+
+function getBetterTypeName(name: string) {
+  // prettier-ignore
+  switch (name) {
+    case 'Plugin$1': return 'rollup.Plugin'
+    case 'TransformResult$1': return 'esbuild_TransformResult'
+    case 'TransformResult$2': return 'rollup.TransformResult'
```

Comment:
I am concerned that these two might suddenly be swapped in some day. For example, if rollup changes its chunking mechanism or renaming algorithm, or our code changes somehow changes the import order.

Reply:
Hmm true. I guess it would be safer if they aren't using namespaces at all, so the worst case is the confusing names.

Maybe I'll need to find another way to handle this, checking import-by-import perhaps then swapping it out 🤔

Reply:
Could this be tested somehow to detect if the order was changed so you can keep the current code?

Reply:
Running a typecheck after building the types could work, but we'd need a cover all the possible JS APIs. I'm currently working on a more strict approach that maybe helps with this.